doc/design/logging.md

```diff
@@ -0,0 +1,402 @@
+   passed tests but output them for failed tests. This keeps the noise low but
+   still provides useful information when needed. Additionally, there shall be an
+   environment variable to circumvent the suppression when desired.
+
+### Solution
+
+#### Class diagram
+
+![logging class diagram](../website/images/logging_classes.svg)
+
+The logger can be customized at compile-time and at runtime. The former is done
+by the `BaseLogger` template parameter and the latter by deriving from the logger.
+
+#### Logging with LogStream
+
+![logging with logstream](../website/images/logging_with_logstream.svg)
+
+#### Logging macro
+
+The `IOX_LOG` macro is intended for general use. It has one parameter which
+sets the log level and delegates `file`, `line` and `function` to the `IOX_LOG_INTERNAL`
+macro. The latter can be used in places like a custom `ASSERT` macro which already
+obtained the `file`, `line` and `function` parameter.
+
+The `IOX_LOG_INTERNAL` uses the `iox::log::internal::SelectedLogStream` struct which is
+specialized for `LogLevel::OFF` to compile to a no-op when the logging data does not
+cause side effects. `IOX_LOG(INFO) << 42;` won't cause side effects but
+`IOX_LOG(INFO) << functionCall();` might have some.
+
+The `IOX_LOG_INTERNAL` calls `self()` on the `LogStream` instance to create an lvalue
+reference to the `LogStream` instance. This eases the implementation of logging
+support for custom types since `IOX_LOG(INFO) << myType;` would require to implement
+an overload with a rvalue `LogStream` reference but `IOX_LOG(INFO) << "#### " << myType;`
+requires a lvalue reference.
+
+#### Behavior before calling Logger::init
+
+In order to have log messages before `Logger::init` is called, the default logger
+is used with `LogLevel::INFO`. It is up to the implementation of the default
+logger what to do with these messages. For iceoryx the default logger is the
+`ConsoleLogger` (this can be changed via the platform abstraction) which will
+print the log messages to the console.
+
+Although, it is possible to use the logger without calling `Logger::init`, this is
+not recommended. This behaviour is only intended to catch important log messages
+from pre-main logger calls.
+
+#### Replacing the logger at runtime
+
+The default console logger can be replaced at runtime to a custom logger which
+can be used to forward the log messages to a logger of another framework.
+
+This is done by deriving from the base `Logger` class and implementing the pure
+virtual `createLogMessageHeader` and `flush` methods. Finally, the derived logger
+must be activated by calling the static `Logger::setActiveLogger` function and
+passing the new logger as argument to the function. The logger must have a static
+lifetime and should therefore be placed in the data segment.
+
+The logger can only be set once and subsequent calls to `Logger::setActiveLogger`
+will not replace the logger anymore and log error messages to the existing logger
```

Comment:
@ithier the `and log error messages ...` is correct since the user needs to be informed that something bad happened. Nevertheless, I rephrased the sentence.

Reply:
@elBoberido OK. With the previous wording I had to read a few times to understand (or maybe I still didn't understand).

Reply:
@ithier to my excuse I'm not a native speaker :sweat_smile: 

What is your understanding of the current wording?

Reply:
@elBoberido The logger won't be replaced AND ADDITIONALLY error messages will be written to the existing logger.

What's not entirely clear is if you mean "a log message indicating an error" will be written if trying to set another logger, or "messages logged by the application with `IOX_ERROR` will keep using the initially set logger.

I would reword it a bit still because it doesn't sound so natural, it is probably nitpicking and not so important so I think it's not worth spending the effort haha.

Reply:
@ithier the old logger remains but an error message will be logged with the old logger and the logger which was supposed to replace the old logger. This is to ensure that all parties get informed about the attempt to replace the logger. How would you phrase it?

Reply:
@elBoberido 

The logger can only be set once and subsequent calls to `Logger::setActiveLogger`
will have no effect. Instead, an error message will be logged to both loggers to notify both parties of the unsuccessful attempt to switch the active logger and the application will continue to use the original logger for subsequent logs.

Reply:
@ithier that's so much better